### PR TITLE
TINY-8334: Updated alloy to allow patching existing DOM elements

### PR DESCRIPTION
**Pull request by lnewson**
Related Ticket: TINY-8334

Description of Changes:

This was a fun one but the core thing we needed from it was to make sure only certain items were re-rendered if a silver dialog spec only changed a child bar/panel components when redialing. So this takes a spike done by @ephox-mogran from a few years ago, which he has then graciously rebased it onto `develop`. So this is largely done by him and then from there I've just taken it, cleaned it up, fixed a few things, made it work in a few more places and added more tests.

The core concept here is that when using `Replacing` or `Reflecting` behaviours alloy will now attempt to patch the existing dom tree instead of recreating it from scratch. This is done by passing in the original element being replaced and then seeing if we are able to patch it (needs to be the same tag, the spec can't have mixed children and the original element can't be used within an existing premade spec). I should add there is a reasonable amount of complexity around `premade` specs as the element can move when you're patching the spec. We've left a fair bit of comments throughout the code that should help to explain this I hope.

Note: I've also fixed a couple test helpers as part of this that were causing headaches when writing the tests. The issue was just that they cleaned up the DOM on failure so you couldn't inspect it.

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/alloy/src/main/ts/ephox/alloy/api/component/GuiFactory.ts [hunk 1]
```diff
@@ -12,16 +12,22 @@ import * as Component from './Component';
...
+      return buildOrPatch(c, Traverse.child(obs, i));
+    })
+  );
 };
 
-const buildFromSpec = (userSpec: SketchSpec): Result<AlloyComponent, string> => {
+const buildFromSpec = (userSpec: SketchSpec, obsoleted: Optional<SugarElement<Node>>): Result<AlloyComponent, string> => {
   const { events: specEvents, ...spec }: SketchSpec = CustomSpec.make(userSpec);
 
   // Build the subcomponents. A spec hierarchy is built from the bottom up.
```
Comment (ephox-mogran) [inline modules/alloy/src/main/ts/ephox/alloy/api/component/GuiFactory.ts:29]: We might want to add some notes to some of these comments explaining briefly what obsoleted is for.

File: modules/alloy/src/main/ts/ephox/alloy/api/component/GuiFactory.ts [hunk 2]
```diff
@@ -94,19 +100,23 @@ const isSketchSpec = (spec: AlloySpec): spec is SketchSpec =>
   Obj.has(spec as SimpleOrSketchSpec, 'uid');
 
 // INVESTIGATE: A better way to provide 'meta-specs'
```
Comment (ephox-mogran) [inline modules/alloy/src/main/ts/ephox/alloy/api/component/GuiFactory.ts:102]: Unfortunately, I have no real idea what "meta-specs" is any more...

File: modules/alloy/src/main/ts/ephox/alloy/behaviour/replacing/ReplaceApis.ts [hunk 3]
```diff
@@ -4,32 +4,42 @@ import { Compare, Insert, SugarElement } from '@ephox/sugar';
...
   Attachment.attachWith(component, child, insertion);
 };
 
+const replace = (component: AlloyComponent, replacee: AlloyComponent, replaceeIndex: number, childSpec: AlloySpec) => {
+  Attachment.detach(replacee);
+  insert(component, (p, c) => Insert.appendAt(p, c, replaceeIndex), childSpec);
+};
+
+const set = (component: AlloyComponent, replaceConfig: ReplacingConfig, replaceState: Stateless, data: AlloySpec[]): void => {
+  // NOTE: we may want to create a behaviour which allows you to switch
```
Comment (ephox-mogran) [inline modules/alloy/src/main/ts/ephox/alloy/behaviour/replacing/ReplaceApis.ts:30]: Redesigning I think was removed over four years ago :) So I clearly kept these comments up-to-date.

File: modules/alloy/src/main/ts/ephox/alloy/behaviour/replacing/ReplacingAll.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import { Traverse } from '@ephox/sugar';
+
+import { AlloyComponent } from '../../api/component/ComponentApi';
+import { AlloySpec } from '../../api/component/SpecTypes';
+import * as AriaFocus from '../../aria/AriaFocus';
+import { patchSpecChildren } from '../../dom/Patching';
+import * as InternalAttachment from '../../system/InternalAttachment';
+
+const withoutReuse = (parent: AlloyComponent, data: AlloySpec[]): void => {
+  // NOTE: we may want to create a behaviour which allows you to switch
```
Comment (ephox-mogran) [inline modules/alloy/src/main/ts/ephox/alloy/behaviour/replacing/ReplacingAll.ts:11]: I think just remove this comment (and the other one)

File: modules/alloy/src/main/ts/ephox/alloy/dom/DomRender.ts [hunk 5]
```diff
@@ -19,13 +22,36 @@ const renderToDom = (definition: DomDefinition.GeneralDefinitionDetail<SugarElem
...
-    // eslint-disable-next-line no-debugger
-    debugger;
+  return subject;
+};
+
+const attemptPatch = (definition: DomDefinition.GeneralDefinitionDetail<SugarElement<Node>>, obsoleted: SugarElement<Element>): Optional<SugarElement<Element>> => {
+  try {
+    const e = reconcileToDom(definition, obsoleted);
+    return Optional.some(e);
+  } catch (err) {
```
Comment (ephox-mogran) [inline modules/alloy/src/main/ts/ephox/alloy/dom/DomRender.ts:32]: It might be worth logging this error.

File: modules/alloy/src/main/ts/ephox/alloy/dom/DomRender.ts [hunk 6]
```diff
@@ -19,13 +22,36 @@ const renderToDom = (definition: DomDefinition.GeneralDefinitionDetail<SugarElem
...
+const hasMixedChildren = (definition: DomDefinition.GeneralDefinitionDetail<SugarElement<Node>>) =>
+  definition.innerHtml.isSome() && definition.domChildren.length > 0;
+
+const renderToDom = (definition: DomDefinition.GeneralDefinitionDetail<SugarElement<Node>>, optObsoleted: Optional<SugarElement<Node>>): SugarElement<Element> => {
+  // If the current tag doesn't match, let's not try to add anything further down the tree.
+  // If it does match though and we don't have mixed children then attempt to patch attributes etc...
+  const canBePatched = (candidate: SugarElement<Node>): candidate is SugarElement<Element> =>
+    SugarNode.name(candidate) === definition.tag && !hasMixedChildren(definition) && !isPremade(candidate);
+
+  const elem = optObsoleted
```
Comment (ephox-mogran) [inline modules/alloy/src/main/ts/ephox/alloy/dom/DomRender.ts:47]: This reads really nicely now.

File: modules/alloy/src/main/ts/ephox/alloy/dom/Reconcile.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    Arr.each(stylesToRemove, (s) => Css.remove(obsoleted, s));
+    Css.setAll(obsoleted, stylesToSet);
+  };
+
+  const existingClasses = Classes.get(obsoleted);
+  const classesToRemove = Arr.difference(existingClasses, definition.classes);
+  const classesToAdd = Arr.difference(definition.classes, existingClasses);
+
+  const updateHtml = (html: string) => {
+    Html.set(obsoleted, html);
```
Comment (ephox-mogran) [inline modules/alloy/src/main/ts/ephox/alloy/dom/Reconcile.ts:44]: Yeah, the original spike has some really strange code in it. I see you've removed that which is good. Hopefully, it wasn't trying to cover some weird edge case. 

Your test coverage is probably pretty good, so it should be fine.

File: modules/alloy/src/main/ts/ephox/alloy/system/InternalAttachment.ts [hunk 8]
```diff
@@ -53,15 +79,52 @@ const replaceChildren = (component: AlloyComponent, newChildren: AlloyComponent[
     } else {
       attach(component, childComp);
     }
-    component.syncComponents();
   });
+  component.syncComponents();
+};
+
+const virtualReplaceChildren = (component: AlloyComponent, newSpecs: AlloySpec[], buildNewChildren: (newSpecs: AlloySpec[]) => AlloyComponent[]): void => {
+  // Detach all existing child components that aren't premade specs
```
Comment (ephox-mogran) [inline modules/alloy/src/main/ts/ephox/alloy/system/InternalAttachment.ts:87]: Maybe a further comment here along the lines of: "we don't want to fire detachedFromDom and attachedToDom again for a premade that has just had its position in the children moved around, so we only detach if initially if we aren't a premade. Premades will be detached later, but only if they are no longer in the child list.

File: modules/alloy/src/test/ts/browser/behaviour/ReplacingTest.ts [hunk 9]
```diff
@@ -1,249 +1,633 @@
...
-        'Initially, has a single span',
-        ApproxStructure.build((s, _str, _arr) => s.element('div', {
-          children: [
-            s.element('span', { })
-          ]
-        })),
-        component.element
-      ),
+    return Arr.map([
+      { comp: withoutReuseComp, label: 'Without reuse' },
```
Comment (ephox-mogran) [inline modules/alloy/src/test/ts/browser/behaviour/ReplacingTest.ts:77]: Nice. Both should work exactly the same way for these tests.

**Review by ephox-mogran — APPROVED** [review]
:shipit:

File: modules/alloy/src/main/ts/ephox/alloy/dom/Reconcile.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  };
+
+  const updateChildren = () => {
+    const children = definition.domChildren;
+    patchDomChildren(obsoleted, children);
+  };
+
+  const updateClasses = () => {
```
Comment (ltrouton) [inline modules/alloy/src/main/ts/ephox/alloy/dom/Reconcile.ts:52]: Just a small thing, would this be better if it was just below the `classes` variables above?

File: modules/alloy/src/test/ts/browser/behaviour/ReplacingTest.ts [hunk 11]
```diff
@@ -1,249 +1,633 @@
...
+          '.replaceAt 2 of 3',
+          spec.comp,
+          [ 'original1', 'original2', 'replaceAt-2' ],
+          [ 'original1', 'original2', 'original3' ],
+          2,
+          Optional.some('replaceAt-2')
+        ),
+
+        sCheckReplaceAt(
+          '.replaceAt 2 with noting',
```
Comment (ltrouton) [inline modules/alloy/src/test/ts/browser/behaviour/ReplacingTest.ts:421]: I think this is a typo?
```suggestion
          '.replaceAt 2 with nothing',
```

**Review by ltrouton — APPROVED** [review]
I don't necessarily understand everything and the intricacies of Alloy. In saying I couldn't see anything that looked off to me so happy to give a tick.

File: modules/alloy/src/main/ts/ephox/alloy/api/component/GuiFactory.ts (hunk 1 above)
Comment (lnewson) [reply modules/alloy/src/main/ts/ephox/alloy/api/component/GuiFactory.ts:29]: Done

File: modules/alloy/src/main/ts/ephox/alloy/behaviour/replacing/ReplaceApis.ts (hunk 3 above)
Comment (lnewson) [reply modules/alloy/src/main/ts/ephox/alloy/behaviour/replacing/ReplaceApis.ts]: Removed this and the other comment that aren't relevant anymore.

File: modules/alloy/src/main/ts/ephox/alloy/dom/DomRender.ts (hunk 5 above)
Comment (lnewson) [reply modules/alloy/src/main/ts/ephox/alloy/dom/DomRender.ts:32]: That might be fine for development but we don't want that spamming in production if it were to happen which is why I didn't include a console log here.

File: modules/alloy/src/main/ts/ephox/alloy/system/InternalAttachment.ts (hunk 8 above)
Comment (lnewson) [reply modules/alloy/src/main/ts/ephox/alloy/system/InternalAttachment.ts]: Added what you had here verbatim basically (minor wording fix)

File: modules/alloy/src/main/ts/ephox/alloy/dom/Reconcile.ts (hunk 10 above)
Comment (lnewson) [reply modules/alloy/src/main/ts/ephox/alloy/dom/Reconcile.ts]: Done

File: modules/alloy/src/test/ts/browser/behaviour/ReplacingTest.ts (hunk 11 above)
Comment (lnewson) [reply modules/alloy/src/test/ts/browser/behaviour/ReplacingTest.ts]: Good catch thanks

File: modules/alloy/src/main/ts/ephox/alloy/dom/Reconcile.ts (hunk 7 above)
Comment (lnewson) [reply modules/alloy/src/main/ts/ephox/alloy/dom/Reconcile.ts:49]: Yeah, all that logic was for handling the case where it had mixed children from what I could tell. That just really can't be patched (or at least not without jumping through a heap of hoops) so I've just instead made mixed children specs unable to be patched (we rarely ever do it and I'd prefer to get rid of being able to do that anyways).
